script/diaspora-dev

```diff
@@ -0,0 +1,434 @@
+  fi
+}
+
+print_usage_full() {
+  # Print overview of available commands
+  # $SCRIPT_NAME [help|-h|--help] leads here
+  echo; echo "<<short description here>>"
+  print_usage_header "$SCRIPT_NAME COMMAND"
+  echo
+  echo "Management Commands:"
+  echo "  setup         Prepare diaspora* to run for development"
+  echo "  start         Start diaspora*"
+  echo "  stop          Stop diaspora*"
+  echo "  restart       Fast restart of diaspora*"
+  echo "  status        Show current instance status of diaspora*"
+  echo "  clean         Reset diaspora* instance"
+  echo
+  echo "Test and Development Commands:"
+  echo "  cucumber      Run cucumber tests"
+  echo "  jasmine       Run jasmine tests"
+  echo "  rspec         Run rspec tests"
+  echo "  migrate       Execute pending migrations"
+  echo
+  echo "Misc. Commands:"
+  echo "  build         Build basic diaspora* environment"
+  echo "  bundle        (Re-)Install gems for diaspora*"
+  echo "  config        Configure diaspora*"
+  echo "  exec          Execute a command in the run environment (advanced)"
+  echo "  help          Show help for commands"
+  echo "  setup-tests   Prepare diaspora* test environment"
+  echo
+  echo "Run '$SCRIPT_NAME help COMMAND' for more information on a command."
+}
+
+
+# ----- Helper functions -----
+
+dia_fetch_upstream() {
+  # Add and fetch upstream develop branch
+  if ! git remote show | grep -q '^upstream$'; then
+    git remote add upstream https://github.com/diaspora/diaspora.git
+  fi
+  git fetch upstream develop
+}
+
+dia_is_configured() {
+  # Check if config files exist
+  [ -f "$DIASPORA_CONFIG_DB" ] && [ -f "$DIASPORA_CONFIG_DIA" ]
+}
+
+exit_if_unconfigured() {
+  # Exit if config does not seem complete
+  dia_is_configured || \
+    echo "Fatal: Config files missing. Run the 'setup' or 'config' command to configure." && \
+    exit 1
+}
+
+dia_is_running() {
+  # Check if diaspora container is running
+  docker-compose ps --services --filter status=running | grep -qx 'diaspora'
+}
+
+dia_is_db_running() {
+  # Check if db container is running
+  docker-compose ps --services --filter status=running | grep -qx $DIASPORA_DOCKER_DB
+}
+
+# ----- Command functions -----
+
+dia_build() {
+  # Build the diaspora Docker container (diaspora:dev-latest)
+  docker-compose build diaspora
+}
+
+dia_bundle() {
+  # Run bundle in order to install all gems into $DIASPORA_ROOT
+  docker-compose run \
+    --rm \
+    --no-deps $1 \
+    diaspora \
+    /bin/sh -c "script/configure_bundler && bin/bundle install --full-index"
+}
+
+dia_clean() {
+  # Delete all containers and volumes
+  docker-compose down -v
+}
+
+dia_config() {
+  # Create rudimentary configuration files if they don't exist
+  [ ! -f "$DIASPORA_ROOT"/public/source.tar.gz ] && touch "$DIASPORA_ROOT"/public/source.tar.gz
+  if [ ! -f "$DIASPORA_CONFIG_DIA" ]; then
+    cp "$DIASPORA_CONFIG_DIA".example "$DIASPORA_CONFIG_DIA"
+  fi
+  if [ ! -f "$DIASPORA_CONFIG_DB" ]; then
+    sed -E '
```

Comment:
From our off-the-record discussions, I know you don't really like this sed, neither do I. I just remembered that ERB will be evaluated within `database.yml`, so you could, in theory, provide a default `database.yml` containing something like

```yml
common: &common
  <<: *<%= ENV["DIASPORA_DOCKER_DB"] %>
```

to easily toggle between the chosen database engine. As the format of `database.yml` is not our invention, but in fact, Rails default behavior, it can be considered pretty stable, and thus I don't have much concerns with having a default `database.yml` shipped with the Docker image.

Reply:
As far as I understand, this would break `script/configure_bundler` as it does not listen to environment variables.

Yes, I am not perfectly happy with this, but it is way better than the rudimentary YAML parser I wrote in awk ...

Reply:
I had the same idea and also suggested that, but if the sed works now, I think we can keep that, it should be as stable as the database.yml :)

Also the `script/configure_bundler` tries to parse the database.yml without ERB, so that wouldn't work at the moment.

I also suggested to create a default database.yml for docker that could be used with sed easier (so same as with ERB, but instead of `<%= ENV["DIASPORA_DOCKER_DB"] %>` use something like `%DIASPORA_DOCKER_DB%` which could be replaced with sed with a oneliner.

So as the sed works now, I think we can keep it, if it creates problems in the future, we can still replace it with an extra docker-specific file. :)

Reply:
Hm, fair points. I am fine with keeping it, if it doesn't cause pain for anyone. :)